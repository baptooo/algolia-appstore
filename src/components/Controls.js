import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import Typography from 'material-ui/Typography';
import Button from 'material-ui/Button';
import Icon from 'material-ui/Icon';
import { withStyles } from 'material-ui/styles';

const styles = theme => ({
  navButton: {
    minWidth: 30
  }
});

const NAV_LENGTH = 7;

const Controls = ({ content, clear, setIndex, setPage, classes }) => {
  const isRatingAsc = content.index === 'appstore';
  const activeStep = content.page;
  const numPages = content.nbPages;
  const navLength = Math.min(numPages, NAV_LENGTH);
  const showNavigation = navLength && numPages > 1;
  const pages = new Array(navLength)
    .fill(0).map((zero, idx) => Math.min(numPages - navLength - 1, activeStep) + idx + 1);

  return (
    <section>
      <Typography variant="headline">
        {content.nbHits} results
        <Button onClick={() => {
          setIndex(isRatingAsc ? 'appstore_orderByRating' : 'appstore');
        }}>
          {isRatingAsc ? (
            <Fragment>
              Rating (DESC)
              <Icon>arrow_drop_down</Icon>
            </Fragment>
          ) : (
            <Fragment>
              Rating (ASC)
              <Icon>arrow_drop_up</Icon>
            </Fragment>
          )}
        </Button>
      </Typography>
      {showNavigation ? (
        <div className={classes.navigation}>
          <Typography variant="caption">Page {activeStep + 1} / {numPages}</Typography>
          <Button className={classes.navButton} size="small" onClick={() => setPage(0)} disabled={activeStep === 0}>
            <Icon>fast_rewind</Icon>
          </Button>
          <Button className={classes.navButton} size="small" onClick={() => setPage(activeStep - 1)} disabled={activeStep === 0}>
            <Icon>navigate_before</Icon>
          </Button>
          {pages.map(page => (
            <Button
              disabled={activeStep === page}
              className={classes.navButton}
              size="small"
              key={`controls-page-${page}`}
              onClick={() => setPage(page)}
            >{page + 1}</Button>
          ))}
          <Button className={classes.navButton} size="small" onClick={() => setPage(activeStep + 1)} disabled={activeStep === numPages - 1}>
            <Icon>navigate_next</Icon>
          </Button>
          <Button className={classes.navButton} size="small" onClick={() => setPage(numPages - 1)} disabled={activeStep === numPages - 1}>
            <Icon>fast_forward</Icon>
          </Button>
        </div>
      ) : null}
      {content.nbHits === 0 ? (
        <Button onClick={() => clear()}>
          Clear <Icon>clear_all</Icon>
        </Button>
      ) : null}
    </section>
  );
}

Controls.propTypes = {
  content: PropTypes.object.isRequired,
  clear: PropTypes.func,
  setIndex: PropTypes.func,
  setPage: PropTypes.func
};

export default withStyles(styles)(Controls);