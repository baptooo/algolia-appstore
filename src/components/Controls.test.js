import React from 'react';
import renderer from 'react-test-renderer';
import Controls from "./Controls";

const mockContent = {
  nbHits: 10,
  hits: new Array(10).fill({}),
  nbPages: 2,
  hitsPerPage: 5,
  page: 0
};

const mockNoContent = {
  nbHits: 0,
};

describe('(Component) Controls', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Controls content={mockContent}/>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should show the clear button when no results', () => {
    const tree = renderer.create(<Controls content={mockNoContent}/>).toJSON();

    expect(tree).toMatchSnapshot();
  });
});