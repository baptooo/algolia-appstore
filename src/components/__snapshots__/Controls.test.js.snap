// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Controls should render correctly 1`] = `
<section>
  <h1
    className="MuiTypography-root-2 MuiTypography-headline-7"
  >
    10
     results
    <button
      className="MuiButtonBase-root-41 MuiButton-root-26"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        Rating (ASC)
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-43"
        >
          arrow_drop_up
        </span>
      </span>
      <span
        className="MuiTouchRipple-root-49"
      />
    </button>
  </h1>
  <div
    className={undefined}
  >
    <span
      className="MuiTypography-root-2 MuiTypography-caption-12"
    >
      Page 
      1
       / 
      2
    </span>
    <button
      className="MuiButtonBase-root-41 MuiButtonBase-disabled-42 MuiButton-root-26 MuiButton-sizeSmall-38 MuiButton-disabled-35 Controls-navButton-1"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-43"
        >
          fast_rewind
        </span>
      </span>
    </button>
    <button
      className="MuiButtonBase-root-41 MuiButtonBase-disabled-42 MuiButton-root-26 MuiButton-sizeSmall-38 MuiButton-disabled-35 Controls-navButton-1"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-43"
        >
          navigate_before
        </span>
      </span>
    </button>
    <button
      className="MuiButtonBase-root-41 MuiButton-root-26 MuiButton-sizeSmall-38 Controls-navButton-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        -3
      </span>
      <span
        className="MuiTouchRipple-root-49"
      />
    </button>
    <button
      className="MuiButtonBase-root-41 MuiButton-root-26 MuiButton-sizeSmall-38 Controls-navButton-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        -2
      </span>
      <span
        className="MuiTouchRipple-root-49"
      />
    </button>
    <button
      className="MuiButtonBase-root-41 MuiButton-root-26 MuiButton-sizeSmall-38 Controls-navButton-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        -1
      </span>
      <span
        className="MuiTouchRipple-root-49"
      />
    </button>
    <button
      className="MuiButtonBase-root-41 MuiButton-root-26 MuiButton-sizeSmall-38 Controls-navButton-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        0
      </span>
      <span
        className="MuiTouchRipple-root-49"
      />
    </button>
    <button
      className="MuiButtonBase-root-41 MuiButton-root-26 MuiButton-sizeSmall-38 Controls-navButton-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        1
      </span>
      <span
        className="MuiTouchRipple-root-49"
      />
    </button>
    <button
      className="MuiButtonBase-root-41 MuiButton-root-26 MuiButton-sizeSmall-38 Controls-navButton-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        2
      </span>
      <span
        className="MuiTouchRipple-root-49"
      />
    </button>
    <button
      className="MuiButtonBase-root-41 MuiButton-root-26 MuiButton-sizeSmall-38 Controls-navButton-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-43"
        >
          navigate_next
        </span>
      </span>
      <span
        className="MuiTouchRipple-root-49"
      />
    </button>
    <button
      className="MuiButtonBase-root-41 MuiButton-root-26 MuiButton-sizeSmall-38 Controls-navButton-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-43"
        >
          fast_forward
        </span>
      </span>
      <span
        className="MuiTouchRipple-root-49"
      />
    </button>
  </div>
</section>
`;

exports[`(Component) Controls should show the clear button when no results 1`] = `
<section>
  <h1
    className="MuiTypography-root-2 MuiTypography-headline-7"
  >
    0
     results
    <button
      className="MuiButtonBase-root-41 MuiButton-root-26"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        Rating (ASC)
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-43"
        >
          arrow_drop_up
        </span>
      </span>
      <span
        className="MuiTouchRipple-root-49"
      />
    </button>
  </h1>
  <button
    className="MuiButtonBase-root-41 MuiButton-root-26"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-27"
    >
      Clear 
      <span
        aria-hidden="true"
        className="material-icons MuiIcon-root-43"
      >
        clear_all
      </span>
    </span>
    <span
      className="MuiTouchRipple-root-49"
    />
  </button>
</section>
`;
